Mathlib/Algebra/Module/DirectLimitAndTensorProduct.lean

```diff
@@ -0,0 +1,141 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import Mathlib.Algebra.DirectLimit
+
+/-!
+Tensor product and direct limits communicate with each other.
+
+-/
+
+open TensorProduct
+
+universe u v w
+
+variable {R : Type u} [CommRing R]
+variable {ι : Type v}
+variable [DecidableEq ι] [Preorder ι]
+variable {G : ι → Type w}
+
+namespace Module
+
+variable [∀ i, AddCommGroup (G i)] [∀ i, Module R (G i)]
+
+variable (f : ∀ i j, i ≤ j → G i →ₗ[R] G j)
+variable (M : Type w) [AddCommGroup M] [Module R M]
+
+/-- limᵢ (G i ⊗ M) -/
+local notation "Lim_G_tensor" =>
+  (DirectLimit (fun i => G i ⊗[R] M) <| fun i j h => TensorProduct.map (f _ _ h) LinearMap.id)
+
+/-- lim G -/
+local notation "Lim_G" => (DirectLimit G f)
+
+noncomputable def directLimitOfTensorProductToTensorProductWithDirectLimit :
+  Lim_G_tensor →ₗ[R] Lim_G ⊗[R] M :=
+DirectLimit.lift _ _ _ _ (fun i => TensorProduct.map (DirectLimit.of _ _ _ _ _) LinearMap.id) <|
+  fun i j h x => x.induction_on (by simp only [map_zero])
+    (fun g m => by simp only [map_tmul, LinearMap.id_coe, id_eq, DirectLimit.of_f])
+    (fun x y hx hy => by
+      dsimp only at hx hy ⊢
+      rw [map_add, map_add, hx, hy, map_add])
+
+variable {M}
+lemma directLimitOfTensorProductToTensorProductWithDirectLimit_apply_of_tmul
+    {i : ι} (g : G i) (m : M) :
+    directLimitOfTensorProductToTensorProductWithDirectLimit f M
+      (DirectLimit.of _ _ _ _ i (g ⊗ₜ m) : Lim_G_tensor) =
+    (DirectLimit.of _ _ _ _ i g : Lim_G) ⊗ₜ m := by
+  dsimp only [directLimitOfTensorProductToTensorProductWithDirectLimit]
+  rw [DirectLimit.lift_of (R := R) (ι := ι) (G := fun i => G i ⊗[R] M)
+        (f := fun i j h => TensorProduct.map (f _ _ h) LinearMap.id) (i := i)
+        (P := Lim_G ⊗[R] M) _ _ (g ⊗ₜ m)]
+  rfl
+
+variable (M)
+noncomputable def tensorProductWithDirectLimitToDirectLimitOfTensorProduct :
+  Lim_G ⊗[R] M →ₗ[R] Lim_G_tensor :=
+TensorProduct.lift <| DirectLimit.lift _ _ _ _
+  (fun i =>
+    { toFun := fun g =>
+      { toFun := fun m => DirectLimit.of R ι (fun i => G i ⊗[R] M)
+          (fun i j h => TensorProduct.map (f _ _ h) LinearMap.id) i (g ⊗ₜ m)
+        map_add' := fun _ _ => by dsimp only; rw [tmul_add, map_add]
+        map_smul' := fun r x => by
+          dsimp only
+          rw [RingHom.id_apply, ← smul_tmul, ← smul_tmul', LinearMap.map_smul] }
```

Comment:
Do we not have this map somewhere else already?

Reply:
I was having trouble finding this map (and the one above this one), if you know where this map is defined please tell me.

Reply:
You can at least write it as composition with `(g ⊗ₜ .)` (`TensorProduct.mk R _ _ g`), right?

Reply:
thanks for the suggestion

Reply:
`rTensor M f`

Reply:
thanks